### PR TITLE
Make `InferCtxtExt::could_impl_trait` more precise, less ICEy

### LINKED ISSUE #119915: ICE: Compiler incorrectly assumes a `Copy` bound is met for a type.
This ICEs on both the latest beta and the latest nightly with the same error.
stable is not affected.

I'm not quite sure what the name of the issue should be so sorry for the bad description.
I tried to comment the code on what seems to be causing the problem however.

### Code

The minimal reproduction code is also here in the playground.
https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=3f34d3bdda874425b1333311501e54e2

```rs
use std::marker::PhantomData;

// The variance of both generics don't seem to matter.
// I tried with *mut E, *const E, and fn(E). Every one gives the same result.
struct Example<E, FakeParam>(PhantomData<(E, FakeParam)>);

 // This can also have a lifetime of any variance, it doesn't appear to matter.
struct NoLifetime;
// Likewise, the variance of `'a` here also doesn't seem to matter.
struct Immutable<'a>(PhantomData<&'a ()>);

// The E: 'a is part of what causes the ICE. if it is just `E`, the compiler correctly
// finds the `Copy` bound missing on `Example<E, Mutable<'a>>`
impl<'a, E: 'a> Copy for Example<E, Immutable<'a>> {}
impl<'a, E: 'a> Clone for Example<E, Immutable<'a>> {
    fn clone(&self) -> Self { *self }
}

impl<E, FakeParam> Example<E, FakeParam> {
    // there is no ICE if this (correctly) takes `&self`.
    unsafe fn change<NewFakeParam>(self) -> Example<E, NewFakeParam> {
        Example(PhantomData)
    }
}

impl<E> Example<E, NoLifetime> {
    fn the_ice(&mut self) -> Example<E, Immutable<'_>> {
        unsafe { self.change() }
    }
}
```

### Meta

`rustc --version --verbose`:

beta:
```
rustc 1.76.0-beta.5 (f732c37b4 2024-01-12)
binary: rustc
commit-hash: f732c37b4175158d3af9e2e156142ffb0bff8969
commit-date: 2024-01-12
host: x86_64-pc-windows-msvc
release: 1.76.0-beta.5
LLVM version: 17.0.6
```
nightly:
```
rustc 1.77.0-nightly (2319be8e2 2024-01-12)
binary: rustc
commit-hash: 2319be8e265dd19973574eb581d28297baf44b11
commit-date: 2024-01-12
host: x86_64-pc-windows-msvc
release: 1.77.0-nightly
LLVM version: 17.0.6
```

### Error output

```
error: internal compiler error: compiler\rustc_middle\src\ty\generic_args.rs:817:13: Unexpected parameter Type(E/#1) when substituting in region 'a (index=0)
thread 'rustc' panicked at /rustc/f732c37b4175158d3af9e2e156142ffb0bff8969\compiler\rustc_errors\src\lib.rs:1119:75:
Box<dyn Any>
```

<details><summary><strong>Backtrace</strong></summary>
<p>

beta backtrace:
[beta-ice-backtrace.txt](https://github.com/rust-lang/rust/files/13926791/beta-ice-backtrace.txt)

nightly backtrace:
[rustc-ice-2024-01-13T04_35_40-20512.txt](https://github.com/rust-lang/rust/files/13926783/rustc-ice-2024-01-13T04_35_40-20512.txt)

</p>
</details>

### PR DESCRIPTION
**Pull request by compiler-errors**
The implementation for `InferCtxtExt::could_impl_trait` was very wrong. Along with being pretty poorly named, way too specific to ADTs, it was also doing impl substitution wrong -- this caused an ICE (#119915).

This PR generalizes that code, gives it a clearer name, makes it stop using the new trait solver (lol), and fixes some fallout bad suggestions that are made worse with the code fix.

Fixes #119915

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 1]
```diff
@@ -1217,19 +1217,22 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
-                                        ty.peel_refs(),
-                                    ),
-                                );
+                                // FIXME: This error message isn't useful, since we're just
+                                // vaguely suggesting to clone a value that already
+                                // implements `Clone`.
+                                //
+                                // A correct suggestion here would take into account the fact
+                                // that inference may be affected by missing types on bindings,
+                                // etc., to improve "tests/ui/borrowck/issue-91206.stderr", for
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:1234]: 91206 is the only test that i consider to have regressed with this PR. the problem is that fixing this code makes it much harder to only tailor the suggestion for when it should apply. I don't think that it's worth it.

File: compiler/rustc_trait_selection/src/infer.rs [hunk 2]
```diff
@@ -86,64 +90,26 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
...
-                    .filter_map(|impl_def_id| {
-                        self.tcx.impl_trait_ref(impl_def_id).map(|r| (impl_def_id, r))
-                    })
-                    .map(|(impl_def_id, imp)| (impl_def_id, imp.skip_binder()))
-                    .find(|(_, imp)| match imp.self_ty().peel_refs().kind() {
-                        ty::Adt(i_def, _) if i_def.did() == def.did() => true,
-                        _ => false,
-                    })
-            {
-                let mut fulfill_cx = FulfillmentCtxt::new(self);
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/infer.rs:109]: Why were we using the new trait solver!?

File: compiler/rustc_trait_selection/src/infer.rs [hunk 3]
```diff
@@ -86,64 +90,26 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
...
-                // We get all obligations from the impl to talk about specific
-                // trait bounds.
-                let obligations = self
-                    .tcx
-                    .predicates_of(impl_def_id)
-                    .instantiate(self.tcx, args)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/infer.rs:115]: Here is the cause of the ICE -- we were trying to substitute an ADT's args into an impl's predicates. There's no guarantee that these generics line up.

File: compiler/rustc_trait_selection/src/infer.rs [hunk 4]
```diff
@@ -86,64 +90,26 @@ impl<'tcx> InferCtxtExt<'tcx> for InferCtxt<'tcx> {
...
 
-    fn could_impl_trait(
+    fn type_implements_trait_shallow(
         &self,
         trait_def_id: DefId,
         ty: Ty<'tcx>,
         param_env: ty::ParamEnv<'tcx>,
     ) -> Option<Vec<traits::FulfillmentError<'tcx>>> {
         self.probe(|_snapshot| {
-            if let ty::Adt(def, args) = ty.kind()
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/infer.rs:96]: Why is this name "could_impl_trait" if it was specific to ADTs? No comment too! :(

**Comment by jackh726** [comment]
Seems like this was added in https://github.com/rust-lang/rust/commit/210a672005aeeb251ebfeeda49c61bb2d44a4480#diff-622a7d238c727ee83660b2a7fe4dbd2c0fd0ac25a432fc531dfdb17c3e516bf1

Using the new trait solver was probably an accident. I'm wondering if it's worth linting against it.

**Comment by compiler-errors** [comment]
I don't think it's worth linting, but I can add assertions against it.

**Comment by jackh726** [comment]
r=me with CI green

cc @estebank

**Comment by compiler-errors** [comment]
(elaborated the comment a bit more, fixed a typo)

**Pull request merged**